ext/nnstreamer/tensor_filter/tensor_filter_mediapipe.cc

```diff
@@ -0,0 +1,358 @@
+#include "mediapipe/framework/formats/image_frame.h"
+#include "mediapipe/framework/formats/image_frame_opencv.h"
+#include "mediapipe/framework/port/commandlineflags.h"
+#include "mediapipe/framework/port/file_helpers.h"
+#include "mediapipe/framework/port/opencv_highgui_inc.h"
+#include "mediapipe/framework/port/opencv_imgproc_inc.h"
+#include "mediapipe/framework/port/opencv_video_inc.h"
+#include "mediapipe/framework/port/parse_text_proto.h"
+#include "mediapipe/framework/port/status.h"
+
+/**
+ * @brief Macro for debug mode.
+ */
+#ifndef DBG
+#define DBG FALSE
+#endif
+
+using nnstreamer::tensor_filter_subplugin;
+
+namespace nnstreamer {
+namespace tensorfilter_mediapipe {
+
+void _init_filter_mediapipe (void) __attribute__ ( (constructor));
+void _fini_filter_mediapipe (void) __attribute__ ( (destructor));
+
+class mediapipe_subplugin final : public tensor_filter_subplugin {
+private:
+  gchar *config_path;
+  size_t frame_timestamp;
+
+  GstTensorsInfo inputInfo;  /**< The tensor info of input tensors */
+  GstTensorsInfo outputInfo;  /**< The tensor info of output tensors */
+
+  mediapipe::CalculatorGraphConfig config;  /**< about .pbtxt file */
+  mediapipe::CalculatorGraph graph;
+
+  int loadMediapipeGraph (const GstTensorFilterProperties * prop);
+  static const char *name;
+  static const accl_hw hw_list[];
+  static const int num_hw = 1;
+  static mediapipe_subplugin *registeredRepresentation;
+
+public:
+  static void init_filter_mediapipe ();
+  static void fini_filter_mediapipe ();
+  static void inputPtrDeleter (uint8_t* a);
+
+  mediapipe_subplugin ();
+  ~mediapipe_subplugin ();
+
+  tensor_filter_subplugin & getEmptyInstance ();
+  void configure_instance (const GstTensorFilterProperties *prop);
+  void invoke (const GstTensorMemory *input, GstTensorMemory *output);
+  void getFrameworkInfo (GstTensorFilterFrameworkInfo &info);
+  int getModelInfo (model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info);
+  int eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data);
+};
+
+const char *mediapipe_subplugin::name = "mediapipe";
+const accl_hw mediapipe_subplugin::hw_list[] = {
+    ACCL_DEFAULT, ACCL_AUTO };
+
+/**
+ * @brief	mediapipe_subplugin Constructor
+ */
+mediapipe_subplugin::mediapipe_subplugin () :
+    tensor_filter_subplugin (),
+    config_path (nullptr)
+{
+  gst_tensors_info_init (&inputInfo);
+  gst_tensors_info_init (&outputInfo);
+
+  frame_timestamp = 0;
+}
+
+/**
+ * @brief	mediapipe_subplugin Destructor
+ */
+mediapipe_subplugin::~mediapipe_subplugin ()
+{
+  mediapipe::Status status;
+
+  g_free (config_path);
+
+  gst_tensors_info_free (&inputInfo);
+  gst_tensors_info_free (&outputInfo);
+
+  for (unsigned int i = 0; i < inputInfo.num_tensors; i++) {
+    status = graph.CloseInputStream (inputInfo.info[i].name);
+    if (!status.ok ()) {
+      std::cerr << "Failed to close input stream" << std::endl;
+    }
+  }
+  status = graph.WaitUntilDone ();
```

Comment:
Q. does this function work without configure/initialization?

Reply:
No, it has to be configured first. Since the configuration is performed right after getting `emptyInstance`, and if it is failed, the `runtime_error` will be called.

Therefore, IMO, we don't need to concern about the NOT configuring case at the destructor.

Reply:
Understood.